classes/task/sendfailednotifications_task.php

```diff
@@ -141,9 +148,31 @@ private function sendmail($msg, $subject, $emailto) {
         $from->email = self::DEFAULT_SENDER_EMAIL;
         $from->deleted = 0;
         $from->mailformat = FORMAT_HTML;
-    
-        $messageid = email_to_user($user, $from, $subject, html_to_text($msg), $msg);
-        return $messageid;
+
+        // if any rows haven't been sent before then set flag to send
+        $rowstosend = false;
+        foreach ($failedrows as $row) {
+            $rv = $DB->get_records("logstore_xapi_notif_sent_log", array("failedlogid" => $row->id, "email" => $emailto));
+            if (empty($rv)) {
+                $rowstosend = true;
+                break;
+            }
+        }
+
+        if ($rowstosend == true) {
+            $messageid = email_to_user($user, $from, $subject, html_to_text($msg), $msg);
+
+            // log that these notifications have been sent
+            $now = time();
+            $rows = array();
+            foreach ($failedrows as $row) {
```

Comment:
I'm wondering if it's necessary to loop these records twice?

Reply:
I was thinking that too but what we are trying to find is if there is a need to send the email to this user because they haven't received any new alerts since the last time it was run. The first iteration is a bit shorter because it exits as soon as a new alert is detected.